@xen-orchestra/vmware-explorer/esxi.mjs

```diff
@@ -0,0 +1,259 @@
+        // don't care about the type for now
+        delete val.attributes
+        // a scalar value : simplify it
+        if (val.$value) {
+          objects[obj.$value][name] = val.$value
+        } else {
+          objects[obj.$value][name] = val
+        }
+      })
+    })
+    return objects
+  }
+
+  async #inspectVmdk(dataStores, currentDataStore, currentPath, filePath) {
+    let diskDataStore, diskPath
+    if (filePath.startsWith('/')) {
+      // disk is on another datastore
+      Object.keys(dataStores).forEach(dataStoreUrl => {
+        if (filePath.startsWith(dataStoreUrl)) {
+          diskDataStore = dataStores[dataStoreUrl].name
+          diskPath = filePath.substring(dataStoreUrl.length + 1)
+        }
+      })
+    } else {
+      diskDataStore = currentDataStore
+      diskPath = currentPath + '/' + filePath
+    }
+    const vmdkRes = await this.download(diskDataStore, diskPath)
+    const text = await vmdkRes.text()
+    const parsed = parseVmdk(text)
+
+    return {
+      ...parsed,
+      datastore: diskDataStore,
+      path: dirname(diskPath),
+      descriptionLabel: ' from esxi'
+    }
+  }
+
+  async getTransferableVmMetadata(vmId) {
+    const search = await this.search('VirtualMachine', ['name', 'config', 'storage', 'runtime', 'snapshot'])
+    if (search[vmId] === undefined) {
+      throw new Error(`VM ${vmId} not found `)
+    }
+    const { config, runtime } = search[vmId]
+
+    const [, dataStore, vmxPath] = config.files.vmPathName.match(/^\[(.*)\] (.+.vmx)$/)
+    const res = await this.download(dataStore, vmxPath)
+    const vmx = parseVmx(await res.text())
+    // list datastores
+    const dataStores = {}
+    Object.values(await this.search('Datastore', ['summary'])).forEach(({ summary }) => {
+      dataStores[summary.url] = summary
+    })
+
+    const disks = []
+
+    for (let scsiIndex = 0; scsiIndex < MAX_SCSI; scsiIndex++) {
+      const scsiChannel = vmx[`scsi${scsiIndex}`]
+      if (scsiChannel === undefined) {
+        continue
+      }
+      for (const diskIndex in Object.values(scsiChannel)) {
+        const disk = scsiChannel[diskIndex]
+        if (typeof disk !== 'object' || disk.deviceType !== 'scsi-hardDisk') {
+          continue
+        }
+        disks.push({
+          ...(await this.#inspectVmdk(dataStores, dataStore, dirname(vmxPath), disk.fileName)),
+          node: `scsi${scsiIndex}:${diskIndex}`,
+        })
+      }
+    }
+    const networks = []
+    for (let ethernetIndex = 0; ethernetIndex < MAX_ETHERNET; ethernetIndex++) {
+      const ethernet = vmx[`ethernet${ethernetIndex}`]
+      if (ethernet === undefined) {
+        continue
+      }
+      networks.push({
+        label: ethernet.networkName,
+        macAddress: ethernet.generatedAddress,
+        isGenerated: ethernet.addressType === 'generated',
+      })
+    }
+    const vmsd = await (await this.download(dataStore, vmxPath.replace(/\.vmx$/, '.vmsd'))).text()
+    let snapshots
+    if(vmsd){
```

Comment:
What happens if the VMSD is missing, do we get a 404 or a 200 with an empty body?